### DISCUSSION TITLE
Task 1: Integrate mpmath as alternative to gymp2

### DISCUSSION THREAD

**Discussion by icmoore** [defipy-devs/defipy#7] [Ideas]
At the moment, the precision handling behind stableswappy utilizes the [gymp2](https://pypi.org/project/gmpy2) package within the [StableswapPoolMath class](https://github.com/defipy-devs/stableswappy/blob/main/python/prod/cst/exchg/StableswapPoolMath.py). The issue is [gymp2](https://pypi.org/project/gmpy2/) has a gplv3 copyleft licence associated with it. The task would be to replace the copyleft licenced python package with the BSD licenced [mpmath](https://mpmath.org/) package.

For DeFiPy to be a true MIT licenced product, it can't be utilizing copyleft software within it's codebase. Technically speaking, its is ok to brand DeFiPy as MIT licensed as the code itself has no issues, however if a company were to want to utilize the codebase, they would be forced to find an alternative to gymp2, as they would be also forced to use the gplv3 licenced gymp2 package, if they wanted to use the stableswap features. 

This task is very algorithmic intensive, and would require a deep understanding of the math behind [Composable Stable Pools](https://docs.balancer.fi/concepts/pools/composable-stable.html). Good tutorials behind composable stable pools can be found [here](https://atulagarwal.dev/posts/curveamm/stableswap/) and [here](https://miguelmota.com/blog/understanding-stableswap-curve/).

**Comment by ABusyHippie** [comment]
[jdugger720](https://github.com/jdugger720) said he is taking a look at this and working on it locally
